### PR TITLE
create network wrapper with basic functionality

### PR DESCRIPTION
**Pull request by ilanDoron**
will enable replacing network contract

### CODE REVIEW

File: contracts/KyberNetworkInterface.sol [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+import "./ERC20Interface.sol";
+import "./KyberReserveInterface.sol";
+
+
+/// @title Kyber Network interface
+contract KyberNetworkInterface {
+    function doTrade(address sender, ERC20 src, uint srcAmount, ERC20 dest, address destAddress, uint maxDestAmount,
+                    uint minConversionRate, address walletId) public payable returns(uint);
+    function getNumReserves() public view returns(uint);
```
Comment (yaronvel) [inline contracts/KyberNetworkInterface.sol]: is `getNumReserves` needed as an interface?

File: contracts/KyberNetwork.sol [hunk 2]
```diff
@@ -13,15 +14,14 @@ import "./FeeBurnerInterface.sol";
...
-contract KyberNetwork is Withdrawable, Utils {
+contract KyberNetwork is KyberNetworkInterface, Withdrawable, Utils {
 
     uint public negligibleRateDiff = 10; // basic rate steps will be in 0.01%
     KyberReserveInterface[] public reserves;
     mapping(address=>bool) public isReserve;
     WhiteListInterface public whiteListContract;
     ExpectedRateInterface public expectedRateContract;
     FeeBurnerInterface    public feeBurnerContract;
-    uint                  public maxGasPrice = 50 * 1000 * 1000 * 1000; // 50 gwei
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: need to discuss. it should be configurable here.

File: contracts/KyberNetwork.sol [hunk 3]
```diff
@@ -347,12 +278,10 @@ contract KyberNetwork is Withdrawable, Utils {
...
     )
-        internal
+        public
+        payable
         returns(uint)
     {
-        require(tx.gasprice <= maxGasPrice);
-        require(validateTradeInput(src, srcAmount, destAddress));
-
         Amount memory amount;
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: do you really need a struct here?

File: contracts/KyberNetwork.sol [hunk 4]
```diff
@@ -475,24 +407,4 @@ contract KyberNetwork is Withdrawable, Utils {
     function calcSrcAmount(ERC20 src, ERC20 dest, uint destAmount, uint rate) internal view returns(uint) {
         return calcSrcQty(destAmount, getDecimals(src), getDecimals(dest), rate);
     }
-
-    /// @notice use token address ETH_TOKEN_ADDRESS for ether
-    /// @dev checks that user sent ether/tokens to contract before trade
-    /// @param src Src token
-    /// @param srcAmount amount of src tokens
-    /// @return true if input is valid
-    function validateTradeInput(ERC20 src, uint srcAmount, address destAddress) internal view returns(bool) {
```
Comment (yaronvel) [inline contracts/KyberNetwork.sol]: I think it should be here, and not in wrapper.

File: contracts/KyberNetworkInterface.sol [hunk 5]
```diff
@@ -0,0 +1,16 @@
+pragma solidity 0.4.18;
+
+
+import "./ERC20Interface.sol";
+import "./KyberReserveInterface.sol";
+
+
+/// @title Kyber Network interface
+contract KyberNetworkInterface {
```
Comment (yaronvel) [inline contracts/KyberNetworkInterface.sol]: max gas price should also be part of the interface.

File: contracts/KyberWrapper.sol [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+import "./KyberReserveInterface.sol";
+import "./KyberNetworkInterface.sol";
+
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////
+/// @title Kyber Network Wrap main contract
+contract KyberWrapper is Withdrawable, Utils {
+
+    KyberNetworkInterface public kyberNetworkContract;
+    uint                  public maxGasPrice = 50 * 1000 * 1000 * 1000; // 50 gwei
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: not sure it should be configurable here. just make a function that returns it from real network.

File: contracts/KyberWrapper.sol [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    bool                  public enabled = false; // network is enabled
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: also not here.

File: contracts/KyberNetworkInterface.sol (hunk 5 above)
Comment (yaronvel) [reply contracts/KyberNetworkInterface.sol]: enabled should also be part of the interface

Comment (yaronvel) [reply contracts/KyberNetworkInterface.sol]: also public info

File: contracts/KyberWrapper.sol [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    mapping(bytes32=>uint) public info; // this is only a UI field for external app.
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: also not here.

File: contracts/KyberWrapper.sol [hunk 9]
```diff
@@ -0,0 +1,161 @@
+pragma solidity 0.4.18;
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: shouldnt we call it KyberNetworkWrapper?

File: contracts/KyberWrapper.sol [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        return actualDestAmount;
+    }
+
+    function setKyberNetworkContract(
+        KyberNetworkInterface _kyberNetworkContract
+    )
+        public
+        onlyAdmin
+    {
+        require(_kyberNetworkContract != address(0));
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: need to log it as an event

File: contracts/KyberWrapper.sol [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        kyberNetworkContract = _kyberNetworkContract;
+    }
+
+    function setMaxGasPrice(uint gasPrice) public onlyAdmin {
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: shouldn't be here

File: contracts/KyberWrapper.sol [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        require(gasPrice > 0);
+        maxGasPrice = gasPrice;
+    }
+
+
+    function setEnable(bool _enable) public onlyAdmin {
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: shouldn't be here

File: contracts/KyberWrapper.sol [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        if (_enable) {
+            require(kyberNetworkContract != address(0));
+        }
+        enabled = _enable;
+    }
+
+    function setInfo(bytes32 field, uint value) public onlyOperator {
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: not here

File: contracts/KyberWrapper.sol [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+        else
+            return token.balanceOf(user);
+    }
+
+    /// @notice use token address ETH_TOKEN_ADDRESS for ether
+    /// @dev checks that user sent ether/tokens to contract before trade
+    /// @param src Src token
+    /// @param srcAmount amount of src tokens
+    /// @return true if input is valid
+    function validateTradeInput(ERC20 src, uint srcAmount, address destAddress) internal view returns(bool) {
```
Comment (yaronvel) [inline contracts/KyberWrapper.sol]: not here

File: contracts/KyberNetwork.sol (hunk 2 above)
Comment (ilanDoron) [reply contracts/KyberNetwork.sol]: it has a set function.

File: contracts/KyberNetwork.sol (hunk 3 above)
Comment (ilanDoron) [reply contracts/KyberNetwork.sol]: yes. 
I get stack too deep compilation error.
u think we can remove any parameter.??

File: contracts/KyberNetwork.sol (hunk 4 above)
Comment (ilanDoron) [reply contracts/KyberNetwork.sol]: OK
will move

**Comment by ilanDoron** [comment]
new PR - number 207 is waiting.

**Pull request closed without merging**